src/stash/src/objects.rs

```diff
@@ -176,6 +176,60 @@ impl From<String> for proto::StringWrapper {
     }
 }
 
+/// Denotes that `Self` is wire compatible with type `T`.
+///
+/// You should not implement this yourself, instead use the `wire_compatible!` macro.
+pub unsafe trait WireCompatible<T> {}
+
+/// Defines one protobuf type as wire compatible with another.
+///
+/// ```text
+/// wire_compatible!(objects_v28::DatabaseKey with objects_v27::DatabaseKey);
+/// ```
+///
+/// Internally this will implement the `WireCompatible<B> for <A>`, e.g.
+/// `WireCompatible<objects_v27::DatabaseKey> for objects_v28::DatabaseKey` and generate `proptest`
+/// cases that will create arbitrary objects of type `B` and assert they can be deserialized with
+/// type `A`, and vice versa.
+macro_rules! wire_compatible {
+    ($a:ident $(:: $a_sub:ident)* with $b:ident $(:: $b_sub:ident)*) => {
+        ::static_assertions::assert_impl_all!(
+            $a $(::$a_sub)* : ::proptest::arbitrary::Arbitrary,
+            ::prost::Message
+        );
+        ::static_assertions::assert_impl_all!(
+            $b $(::$b_sub)*  : ::proptest::arbitrary::Arbitrary,
+            ::prost::Message
+        );
+
+        unsafe impl $crate::objects::WireCompatible< $b $(::$b_sub)* > for $a $(::$a_sub)* {}
+        unsafe impl $crate::objects::WireCompatible< $a $(::$a_sub)* > for $b $(::$b_sub)* {}
+
+        ::paste::paste! {
+            ::proptest::proptest! {
+                #[mz_ore::test]
+                #[cfg_attr(miri, ignore)] // slow
+                fn [<proptest_wire_compat_ $a:snake $(_$a_sub:snake)* _to_ $b:snake $(_$b_sub:snake)* >](a: $a $(::$a_sub)* ) {
+                    use ::prost::Message;
+                    let bytes = a.encode_to_vec();
+                    let decoded = $b $(::$b_sub)*::decode(&bytes[..]);
+                    ::proptest::prelude::prop_assert!(decoded.is_ok());
+                }
+
+                #[mz_ore::test]
+                #[cfg_attr(miri, ignore)] // slow
+                fn [<proptest_wire_compat_ $b:snake $(_$b_sub:snake)* _to_ $a:snake $(_$a_sub:snake)* >](b: $b $(::$b_sub)* ) {
+                    use ::prost::Message;
+                    let bytes = b.encode_to_vec();
+                    let decoded = $a $(::$a_sub)*::decode(&bytes[..]);
+                    ::proptest::prelude::prop_assert!(decoded.is_ok());
+                }
+            }
```

Comment:
Is is worth adding a test that does something like `assert_eq!(a.encode_to_vec(), b.encode_to_vec())`

Reply:
Good thought, I added the assert!

FWIW I'm pretty sure this is true for protobuf, but there might be an edge case where this assert would fail, meanwhile the types are wire compatible. It definitely makes the test more strict though which IMO is a good idea, and we can relax the constraint in the future if need be!